osu.Game/Overlays/Chat/ChannelList/ChannelList.cs

```diff
@@ -0,0 +1,175 @@
+using osu.Game.Graphics.Sprites;
+using osu.Game.Online.Chat;
+using osuTK;
+
+namespace osu.Game.Overlays.Chat.ChannelList
+{
+    public class ChannelList : Container
+    {
+        public Action<Channel>? OnRequestSelect;
+        public Action<Channel>? OnRequestLeave;
+
+        public readonly BindableBool SelectorActive = new BindableBool();
+
+        private readonly Dictionary<Channel, ChannelListItem> channelMap = new Dictionary<Channel, ChannelListItem>();
+
+        private ChannelListItemFlow publicChannelFlow = null!;
+        private ChannelListItemFlow privateChannelFlow = null!;
+
+        [BackgroundDependencyLoader]
+        private void load(OverlayColourProvider colourProvider)
+        {
+            Children = new Drawable[]
+            {
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colourProvider.Background6,
+                },
+                new ChannelListScrollContainer
+                {
+                    Padding = new MarginPadding { Vertical = 7 },
+                    RelativeSizeAxes = Axes.Both,
+                    ScrollbarAnchor = Anchor.TopLeft,
+                    ScrollDistance = 35f,
+                    Child = new FillFlowContainer
+                    {
+                        Direction = FillDirection.Vertical,
+                        RelativeSizeAxes = Axes.X,
+                        AutoSizeAxes = Axes.Y,
+                        Children = new Drawable[]
+                        {
+                            publicChannelFlow = new ChannelListItemFlow("CHANNELS"),
+                            new ChannelListSelector
+                            {
+                                Margin = new MarginPadding { Bottom = 10 },
+                                SelectorActive = { BindTarget = SelectorActive },
+                            },
+                            privateChannelFlow = new ChannelListItemFlow("DIRECT MESSAGES"),
+                        },
+                    },
+                },
+            };
+        }
+
+        public void AddChannel(Channel channel)
+        {
+            if (channelMap.ContainsKey(channel))
+                return;
+
+            ChannelListItem item = new ChannelListItem(channel);
+            item.OnRequestSelect += chan => OnRequestSelect?.Invoke(chan);
+            item.OnRequestLeave += chan => OnRequestLeave?.Invoke(chan);
+
+            ChannelListItemFlow flow = getFlowForChannel(channel);
+            channelMap.Add(channel, item);
+            flow.Add(item);
+        }
+
+        public void RemoveChannel(Channel channel)
+        {
+            if (!channelMap.ContainsKey(channel))
+                return;
+
+            ChannelListItem item = channelMap[channel];
+            ChannelListItemFlow flow = getFlowForChannel(channel);
+
+            channelMap.Remove(channel);
+            flow.Remove(item);
+            item.Expire();
```

Comment:
`Expire()` does essentially nothing on an item that has already been removed from the hierarchy (which, in this case, it has been with the preceding case). The way expiry works is by setting lifetime on a drawable and letting the next update of the composite drawable containing it clean it up.

It also slightly worries me that `item`'s subscriptions to `OnRequest{Select,Remove}` are not cleaned up on removal. I'm not sure if it will lead to a leak, but it may.

Reply:
I think a leak only occurs if the event publisher (the `ChannelListItem` in this case) lives longer than the event subscriber (the parent `ChannelList`) so I think this is fine.